vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,138 @@
+                 ctx: xgr.CompiledGrammar) -> None:
+        # https://xgrammar.mlc.ai/docs/api/python/index.html#xgrammar.GrammarMatcher.find_jump_forward_string for jump-forward decoding
+        # TODO: support max_rollback_tokens
+        self.matcher = matcher
+        self.vocab_size = vocab_size
+        self.ctx = ctx
+        self.prefilled = False
+
+    def accept_token(self, token: int) -> bool:
+        # NOTE: accept_token will determines whether we accept this token
+        # and will also update the machine state
+        return self.matcher.accept_token(token)
+
+    def allocate_bitmask(self, batch_size: int,
+                         vocab_size: int) -> torch.Tensor:
+        return xgr.allocate_token_bitmask(batch_size, vocab_size)
+
+    # this should be ran in parallel with model decoding
+    def fill_bitmask(self, bitmask: torch.Tensor, idx: int) -> None:
+        self.matcher.fill_next_token_bitmask(bitmask, idx)
+
+    @staticmethod
+    def apply_bitmask(logits: torch.Tensor, vocab_mask: torch.Tensor) -> None:
+        xgr.apply_token_bitmask_inplace(logits, vocab_mask)
+
+    def reset(self):
+        self.matcher.reset()
+
+    def copy(self):
+        return Grammar(matcher=xgr.GrammarMatcher(self.ctx),
+                       vocab_size=self.vocab_size,
+                       ctx=self.ctx)
+
+    def __copy__(self):
+        return self.copy()
+
+
+@dataclass
+class GrammarCache:
+    value: Optional[Grammar]
+    event: threading.Event
+
+
+class GuidedDecodingManager:
+
+    def __init__(self, vllm_config: VllmConfig):
+        tokenizer_group = init_tokenizer_from_configs(
+            model_config=vllm_config.model_config,
+            scheduler_config=vllm_config.scheduler_config,
+            parallel_config=vllm_config.parallel_config,
+            lora_config=vllm_config.lora_config)
+        tokenizer_group.ping()
+        self.model_config = vllm_config.model_config
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
+        self.tokenizer = tokenizer_group.get_lora_tokenizer(None)
+        self.grammar_cache: dict[GuidedDecodingKey, GrammarCache] = {}
+        self.executor = ThreadPoolExecutor()
+        self._lock = threading.Lock()
+
+    def initialize_cache(self, key: GuidedDecodingKey) -> Grammar:
+        request_type, grammar_spec = key
+        tokenizer_info = xgr.TokenizerInfo.from_huggingface(
+            self.tokenizer, vocab_size=self.vocab_size)
+        compiler = xgr.GrammarCompiler(tokenizer_info, max_threads=8)
+
+        if request_type == "json":
+            if type(grammar_spec) is not str:
+                ctx = compiler.compile_builtin_json_grammar()
+            else:
+                ctx = compiler.compile_json_schema(grammar_spec)
+        elif request_type == "grammar":
+            ctx = compiler.compile_grammar(grammar_spec)
+        else:
+            raise ValueError("grammar is not of valid supported types.")
```

Comment:
Print the unsupported `request_type` in the error

Reply:
addressed